### PR TITLE
Handle RelayState in preparing a SAMLAuthN Request

### LINKED ISSUE #46232: SAML `_security/saml/prepare` API should allow consumer to specify `RelayState`
Currently Kibana heavily relies on URL fragments to navigate between the apps and store applied filters. And since URL fragments are never sent to the server Kibana didn't preserve them during SAML handshake. Moreover URL fragments used in Kibana can be pretty large so we can't store them in the session cookie like we do for SAML request ID. We're planning to change this in 7.5 (see https://github.com/elastic/kibana/pull/44513) and store full URL in the `RelayState` query string parameter we send to IdP and receive back once user is authenticated.

As we discussed with @jkakavas in https://github.com/elastic/kibana/issues/18392#issuecomment-525192963 it'd make sense for `_security/saml/prepare` API to accept the arbitrary string as an additional argument with  that will eventually be used as `RelayState` query string parameter in the IdP redirect URL this API returns.

cc @kobelb

### PR DESCRIPTION
**Pull request by jkakavas**
This change allows for the caller of the `saml/prepare` API to pass
a `relay_state` parameter that will then be part of the redirect
URL in the response as the `RelayState` query parameter.

The SAML IdP is required to reflect back the value of that relay
state when sending a SAML Response. The caller of the APIs can
then, when receiving the SAML Response, read and consume the value
as it see fit.

Resolves: https://github.com/elastic/elasticsearch/issues/46232

### CODE REVIEW

**Comment by tvernum** [comment]
At first glance this looks fine, but let's wait and see how Kibana wants to deal with the length limit on `RelayState` before we commit the change.

**Comment by jkakavas** [comment]
I agree to wait for Kibana but I think the possibility for the caller of the API to set the relaystate we send with the authentication request is useful to have ( Since we do already handle the possibility of a relay state parameter internally ) even if kibana doesn't use this now for passing the target url

**Comment by jkakavas** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlPrepareAuthenticationRequest.java [hunk 1]
```diff
@@ -67,5 +83,8 @@ public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         out.writeOptionalString(realmName);
         out.writeOptionalString(assertionConsumerServiceURL);
+        if (out.getVersion().onOrAfter(Version.V_7_5_0)){
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlPrepareAuthenticationRequest.java]:
```suggestion
        if (out.getVersion().onOrAfter(Version.V_7_5_0)) {
```

**Review by tvernum — APPROVED** [review]

**Pull request merged**